### Will a Customer Accept the Coupon?

**Context**

Imagine driving through town and a coupon is delivered to your cell phone for a restaraunt near where you are driving. Would you accept that coupon and take a short detour to the restaraunt? Would you accept the coupon but use it on a sunbsequent trip? Would you ignore the coupon entirely? What if the coupon was for a bar instead of a restaraunt? What about a coffee house? Would you accept a bar coupon with a minor passenger in the car? What about if it was just you and your partner in the car? Would weather impact the rate of acceptance? What about the time of day?

Obviously, proximity to the business is a factor on whether the coupon is delivered to the driver or not, but what are the factors that determine whether a driver accepts the coupon once it is delivered to them? How would you determine whether a driver is likely to accept a coupon?

**Overview**

The goal of this project is to use what you know about visualizations and probability distributions to distinguish between customers who accepted a driving coupon versus those that did not.

**Data**

This data comes to us from the UCI Machine Learning repository and was collected via a survey on Amazon Mechanical Turk. The survey describes different driving scenarios including the destination, current time, weather, passenger, etc., and then ask the person whether he will accept the coupon if he is the driver. Answers that the user will drive there ‘right away’ or ‘later before the coupon expires’ are labeled as ‘Y = 1’ and answers ‘no, I do not want the coupon’ are labeled as ‘Y = 0’.  There are five different types of coupons -- less expensive restaurants (under \\$20), coffee houses, carry out & take away, bar, and more expensive restaurants (\\$20 - \\$50). 

**Deliverables**

Your final product should be a brief report that highlights the differences between customers who did and did not accept the coupons.  To explore the data you will utilize your knowledge of plotting, statistical summaries, and visualization using Python. You will publish your findings in a public facing github repository as your first portfolio piece. 





### Data Description
Keep in mind that these values mentioned below are average values.

The attributes of this data set include:
1. User attributes
    -  Gender: male, female
    -  Age: below 21, 21 to 25, 26 to 30, etc.
    -  Marital Status: single, married partner, unmarried partner, or widowed
    -  Number of children: 0, 1, or more than 1
    -  Education: high school, bachelors degree, associates degree, or graduate degree
    -  Occupation: architecture & engineering, business & financial, etc.
    -  Annual income: less than \\$12500, \\$12500 - \\$24999, \\$25000 - \\$37499, etc.
    -  Number of times that he/she goes to a bar: 0, less than 1, 1 to 3, 4 to 8 or greater than 8
    -  Number of times that he/she buys takeaway food: 0, less than 1, 1 to 3, 4 to 8 or greater
    than 8
    -  Number of times that he/she goes to a coffee house: 0, less than 1, 1 to 3, 4 to 8 or
    greater than 8
    -  Number of times that he/she eats at a restaurant with average expense less than \\$20 per
    person: 0, less than 1, 1 to 3, 4 to 8 or greater than 8
    -  Number of times that he/she goes to a bar: 0, less than 1, 1 to 3, 4 to 8 or greater than 8
    

2. Contextual attributes
    - Driving destination: home, work, or no urgent destination
    - Location of user, coupon and destination: we provide a map to show the geographical
    location of the user, destination, and the venue, and we mark the distance between each
    two places with time of driving. The user can see whether the venue is in the same
    direction as the destination.
    - Weather: sunny, rainy, or snowy
    - Temperature: 30F, 55F, or 80F
    - Time: 10AM, 2PM, or 6PM
    - Passenger: alone, partner, kid(s), or friend(s)


3. Coupon attributes
    - time before it expires: 2 hours or one day

In [1]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
import missingno as msno
import scipy

### Problems

Use the prompts below to get started with your data analysis.  

1. Read in the `coupons.csv` file.




In [2]:
data = pd.read_csv('data/coupons.csv')

In [3]:
data.head()

destination  passanger weather  temperature  time  \
0  No Urgent Place      Alone   Sunny           55   2PM   
1  No Urgent Place  Friend(s)   Sunny           80  10AM   
2  No Urgent Place  Friend(s)   Sunny           80  10AM   
3  No Urgent Place  Friend(s)   Sunny           80   2PM   
4  No Urgent Place  Friend(s)   Sunny           80   2PM   

                  coupon expiration  gender age      maritalStatus  ...  \
0        Restaurant(<20)         1d  Female  21  Unmarried partner  ...   
1           Coffee House         2h  Female  21  Unmarried partner  ...   
2  Carry out & Take away         2h  Female  21  Unmarried partner  ...   
3           Coffee House         2h  Female  21  Unmarried partner  ...   
4           Coffee House         1d  Female  21  Unmarried partner  ...   

   CoffeeHouse CarryAway RestaurantLessThan20 Restaurant20To50  \
0        never       NaN                  4~8              1~3   
1        never       NaN                  4~8              1~3   
2        never       NaN                  4~8              1~3   
3        never       NaN                  4~8              1~3   
4        never       NaN                  4~8              1~3   

  toCoupon_GEQ5min toCoupon_GEQ15min toCoupon_GEQ25min direction_same  \
0                1                 0                 0              0   
1                1                 0                 0              0   
2                1                 1                 0              0   
3                1                 1                 0              0   
4                1                 1                 0              0   

  direction_opp  Y  
0             1  1  
1             1  0  
2             1  1  
3             1  0  
4             1  0  

[5 rows x 26 columns]

2. Investigate the dataset for missing or problematic data.

#### Data types and non-null count

In [4]:
#Find out data types & number of Null values
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12684 entries, 0 to 12683
Data columns (total 26 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   destination           12684 non-null  object
 1   passanger             12684 non-null  object
 2   weather               12684 non-null  object
 3   temperature           12684 non-null  int64 
 4   time                  12684 non-null  object
 5   coupon                12684 non-null  object
 6   expiration            12684 non-null  object
 7   gender                12684 non-null  object
 8   age                   12684 non-null  object
 9   maritalStatus         12684 non-null  object
 10  has_children          12684 non-null  int64 
 11  education             12684 non-null  object
 12  occupation            12684 non-null  object
 13  income                12684 non-null  object
 14  car                   108 non-null    object
 15  Bar                   12577 non-null

Following columns have Null values: car, Bar, CoffeeHouse, CarryAway, RestaurantLessThan20, Restaurant20To50 

#### Number of null values

In [5]:
data.isna().sum()

destination                 0
passanger                   0
weather                     0
temperature                 0
time                        0
coupon                      0
expiration                  0
gender                      0
age                         0
maritalStatus               0
has_children                0
education                   0
occupation                  0
income                      0
car                     12576
Bar                       107
CoffeeHouse               217
CarryAway                 151
RestaurantLessThan20      130
Restaurant20To50          189
toCoupon_GEQ5min            0
toCoupon_GEQ15min           0
toCoupon_GEQ25min           0
direction_same              0
direction_opp               0
Y                           0
dtype: int64

#### Unique value counts of each columns to discover problematic data

In [ ]:
#Check value counts for unusual values
data['destination'].value_counts(dropna=False)

In [ ]:
data['passanger'].value_counts(dropna=False)

In [ ]:
data['weather'].value_counts(dropna=False)

In [ ]:
data['temperature'].value_counts(dropna=False)

In [ ]:
data['time'].value_counts(dropna=False)

In [ ]:
data['coupon'].value_counts(dropna=False)

In [ ]:
data['expiration'].value_counts(dropna=False)

In [ ]:
data['gender'].value_counts(dropna=False)

In [ ]:
data['age'].value_counts(dropna=False)

In [ ]:
data['maritalStatus'].value_counts(dropna=False)

In [ ]:
data['has_children'].value_counts(dropna=False)

In [ ]:
data['education'].value_counts(dropna=False)

In [ ]:
data['occupation'].value_counts()

In [ ]:
data['income'].value_counts(dropna=False)

In [ ]:
data['car'].value_counts(dropna=False)

In [ ]:
#feature meaning: how many times do you go to a bar every month?
data['Bar'].value_counts(dropna=False)

In [ ]:
#feature meaning: how many times do you go to a coffeehouse every month?
data['CoffeeHouse'].value_counts(dropna=False)

In [ ]:
#feature meaning: how many times do you get take-away food every month?
data['CarryAway'].value_counts(dropna=False)

In [ ]:
#feature meaning: how many times do you go to a restaurant with an average expense per person of less than $20 every month?
data['RestaurantLessThan20'].value_counts(dropna=False)

In [ ]:
#feature meaning: how many times do you go to a restaurant with average expense per person of $20 - $50 every month?
data['Restaurant20To50'].value_counts(dropna=False)

In [ ]:
#feature meaning: driving distance to the restaurant/bar for using the coupon is greater than 5 minutes
data['toCoupon_GEQ5min'].value_counts(dropna=False)

In [ ]:
#feature meaning: driving distance to the restaurant/bar for using the coupon is greater than 15 minutes
data['toCoupon_GEQ15min'].value_counts(dropna=False)

In [ ]:
#feature meaning: driving distance to the restaurant/bar for using the coupon is greater than 25 minutes
data['toCoupon_GEQ25min'].value_counts(dropna=False)

In [ ]:
#feature meaning: whether the restaurant/bar is in the same direction as your current destination
data['direction_same'].value_counts(dropna=False)

In [ ]:
#feature meaning: whether the restaurant/bar is in the opposite direction as your current destination
data['direction_opp'].value_counts(dropna=False)

In [ ]:
data['Y'].value_counts(dropna=False)

* After examining missing values and possible poblometic values, it is found that the 'car' variable only has 108 non-null values out of 12684 records. Also, 'toCoupon_GEQ5min' varibale has the same values for all the records, so it does not add any additional information for the analysis. 

* Also, it's a minor thing, but since the column name 'passanger' is a typo of the word, passenger, I changed the column name.
* Some values in the passenger column contains '(s)'. I think it is unnecessory becuase we don't have any information about the number of passengers, so I removed '(s)'.

In [ ]:
#Change the column name 'passanger' to 'passenger'
data.rename(columns = {'passanger':'passenger'}, inplace = True)

In [ ]:
#Remove '(s)' in the values of the passenger column
import warnings
warnings.filterwarnings('ignore')
data['passenger'] = data['passenger'].str.replace("\(s\)", "")

3. Decide what to do about your missing data -- drop, replace, other...

#### car
More than 99% of the values are missing. We will drop the column.

In [ ]:
#Drop 'car' column
data.drop(columns=['car'], inplace=True)

#### toCoupon_GEQ5min
'toCoupon_GEQ5min' varibale has the same values for all the records. Since it does not provide additional information to the dataset, we will also drop this column.


In [ ]:
#Drop 'toCoupon_GEQ5min' column
data.drop(columns=['toCoupon_GEQ5min'], inplace=True)

#### direction_opp
'direction_opp' varibale has the exact opposite values to the 'direction_same' variable. We can delete the column since it does not provide additional information to the dataset.


In [ ]:
#Drop 'direction_opp' column
data.drop(columns=['direction_opp'], inplace=True)

#### Other columns with missing data
There are five more columns with missing data: Bar, CoffeeHouse, CarryAway, RestaurantLessThan20, Restaurant20To50
There are different types of missing data, and we need to figure out if missing data are related to each other before we decide what to do with them.

* MCAR : Missing completely at random
* MAR: Missing at random
* MNAR: Missing not at random

Knowing this mechanism is important becuase, according to [this article](https://www.theanalysisfactor.com/missing-data-mechanism/), multiple imputation and maximum likelihood assume the data are at least missing at random, and deletion requires the data are MCAR in order to not introduce bias in the results.

To explore the missing values in the data, I used 'missingno' library.


In [ ]:
#Matrix Chart
na_columns = ['Bar', 'CoffeeHouse', 'CarryAway', 'RestaurantLessThan20', 'Restaurant20To50']
na_data = data[na_columns]
msno.matrix(na_data, figsize=(10, 12))
plt.title("Missing Data Matrix Plot")
plt.show()

This matrix plot shows how the missing data is distributed in the dataset. The gray area is where the data is present, and the white area is where the data is missing. It seems that the missing values in each column are not related to each other. To delve into the relationship, let's look at the heatmap.

In [ ]:
#Heatmap
msno.heatmap(na_data)
plt.title('Missing Data Heatmap')
plt.show()

Heatmap shows the relationship of the presence of null values between each of the different columns. If the correlation value is close to positive 1, it means that the presence of null values of one column is positively correlated to the presence of the other column. The low correlation again shows that the missing data is at least MAR (missing at random) and that we can imputate the data. 

#### Data Imputation
I found that there is low correlation of the presence of missing values between each column that has null values. Since all columns with missing data are captegorical variable, I will impute the missing data with the most frequent value. 

In [ ]:
clean_data = data

for x in clean_data.columns[clean_data.isna().any()]:
    clean_data = clean_data.fillna({x: clean_data[x].value_counts().idxmax()})
    
clean_data.head(10)

In [ ]:
#Check if there is any null value
clean_data.info()

All null-values are replaced by the most frequent vlues. 

4. What proportion of the total observations chose to accept the coupon? 



In [ ]:
clean_data.query('Y == 1').shape[0] / clean_data.shape[0]

56.84% of the observed data accepted the coupn.

5. Use a bar plot to visualize the `coupon` column.

First, let's visualize the count of each value in the coupon column.

In [ ]:
import seaborn as sns
sns.set_theme(style="whitegrid")
sns.countplot(data=clean_data, x="coupon")
plt.title("Histogram of the types of coupons")
plt.xlabel('Types of coupons')
plt.ylabel('Count')
plt.xticks(rotation=45)
plt.show()

I also visualized the number of coupons that were accepted by the drivers right next to the ones that were not accepted.

In [ ]:
sns.countplot(data=clean_data, x="coupon", hue="Y")
plt.title("Count of the coupon acceptance by coupon type")
plt.xlabel('Types of coupons')
plt.ylabel('Count')
plt.xticks(rotation=45)

plt.legend(title='Acceptance', labels=['Not used', 'Accepted'])
plt.show()

This bar plot below shows what percentage of the drivers accepted the coupons in each coupon category.

In [ ]:
#Use barplot to visualze the acceptance rate of each coupon type
grouped = clean_data.groupby('coupon')[['Y']].mean()
grouped = grouped.reset_index()
sns.barplot(data=grouped, x='coupon', y='Y')
plt.title("Acceptance rate by coupon type")
plt.xlabel('Types of coupons')
plt.ylabel('Acceptance Rate')
plt.xticks(rotation=45)
plt.show()

It shows that 'Carry out & Take away' coupons and 'Restauran(<20)' coupons got accepted more compared to other types of coupons. 

6. Use a histogram to visualize the temperature column.

In [ ]:
sns.histplot(data=clean_data, x="temperature")
plt.title("Histogram of the temperature variable")
plt.xlabel('Temperature')
plt.ylabel('Count')
plt.show()

Even though the temperature variable contains numerical values, it does not  contain continuous values. There are only three unique variables, and it functions almost like a categorical variable with ordinal characteristics. So, I think in this case countplot captures the better picture rather than the usual histogram.

In [ ]:
sns.countplot(data=clean_data, x="temperature")
plt.title("Countplot of the temperature variable")
plt.xlabel('Temperature')
plt.ylabel('Count')
plt.show()

**Investigating the Bar Coupons**

Now, we will lead you through an exploration of just the bar related coupons.  

1. Create a new `DataFrame` that contains just the bar coupons.


In [ ]:
#Create bar_data from the cleaned data
bar_data = clean_data.query('coupon == "Bar"')
bar_data

2. What proportion of bar coupons were accepted?


In [ ]:
bar_data.query('Y == 1').shape[0]/bar_data.shape[0]

41.00% of the bar coupons were accepted.

3. Compare the acceptance rate between those who went to a bar 3 or fewer times a month to those who went more.


* Note that I created the bar_data from the "cleaned" dataset which imputated the missing data of Bar, CoffeeHouse, CarryAway, RestaurantLessThan20, Restaurant20To50 columns with the most frequent values

In [ ]:
ls_and_equal3 = ['1~3','less1','never']
gt3 = ['4~8','gt8']

acceptance_rate_ls3 = bar_data.query('Bar in @ls_and_equal3')['Y'].mean()
acceptance_rate_gt3 = bar_data.query('Bar in @gt3')['Y'].mean()


sns.barplot(x=['3 or fewer','More than 3'], y=[acceptance_rate_ls3, acceptance_rate_gt3])
plt.xlabel('Frequency of bar visits in a month')
plt.ylabel('Acceptance Rate')
plt.title('Bar Coupon Acceptance Rates by Bar Visit Frequency')
plt.show()


In [ ]:
print("Acceptance rate of bar coupons by drivers who visit bars 3 or fewer times:", round(acceptance_rate_ls3 *100, 2),"%")
print("Acceptance rate of bar coupons by drivers who visit bars more than 3 times:", round(acceptance_rate_gt3 *100, 2),"%")

Drivers who go to the bar more 3 times a month show higher acceptance rate for bar coupons than those who visit bars less often.

4. Compare the acceptance rate between drivers who go to a bar more than once a month and are over the age of 25 to the all others.  Is there a difference?


In [ ]:
gt_and_equal1 = ['1~3','4~8','gt8']
lsl = ['less1','never']
age_gt25 = ['26', '31', '36', '41', '46', '50plus']
age_ls25 = ['21', 'below21']

acceptance_rate_gt1_gt25 = bar_data.query('Bar in @gt_and_equal1 and age in @age_gt25')['Y'].mean()
acceptance_rate_ls1_ls25 = bar_data.query('Bar in @lsl or age in @age_ls25')['Y'].mean()

sns.barplot(x=['Bar visit >= 1 per month & age > 25','All others'], y=[acceptance_rate_gt1_gt25, acceptance_rate_ls1_ls25])
plt.xlabel('Frequency of bar visits in a month & Age')
plt.ylabel('Acceptance Rate')
plt.title('Bar Coupon Acceptance Rates by Bar Visit Frequency & Age')
plt.show()

In [ ]:
print("Acceptance rate of bar coupons by drivers who visit bars more than once a month and whose age is older than 25:", round(acceptance_rate_gt1_gt25 *100, 2),"%")
print("Acceptance rate of bar coupons by all other drivers:", round(acceptance_rate_ls1_ls25 *100, 2),"%")

There is a clear difference in the acceptance rate between the two groups. Drivers who visit bars more than once a month and older than 25 years old show higher acceptance rate for bar coupons than all other drivers.

5. Use the same process to compare the acceptance rate between drivers who go to bars more than once a month and had passengers that were not a kid and had occupations other than farming, fishing, or forestry. 


I interpreted this question asking to "use the same process" as to **compare the specified group to all others**.

In [ ]:
gt_and_equal1 = ['1~3','4~8','gt8']
lsl = ['less1','never']
passenger_nokid = ['Friend', 'Partner']
group_criteria= bar_data.query('Bar in @gt_and_equal1 and passenger in @passenger_nokid and occupation !="Farming Fishing & Forestry"')

#Check if the querying is done correctly
#print(group_criteria['passenger'].value_counts(dropna=False))
#print(group_criteria['occupation'].value_counts(dropna=False))


In [ ]:
acceptance_rate_gt1_nokid_nofff = bar_data.query('Bar in @gt_and_equal1 and passenger in @passenger_nokid and occupation !="Farming Fishing & Forestry"')['Y'].mean()
acceptance_rate_all_other = bar_data.query('Bar in @lsl or passenger not in @passenger_nokid or occupation =="Farming Fishing & Forestry"')['Y'].mean()

sns.barplot(x=['Bar visit >= 1/month & non-kid passenger & no Farming Fishing Forestry occupation','All others'], y=[acceptance_rate_gt1_nokid_nofff, acceptance_rate_all_other])
plt.xlabel('Frequency of bar visits in a month & Passenger & Occupation')
plt.ylabel('Acceptance Rate')
plt.title('Bar Coupon Acceptance Rates by Bar Visit Frequency & Passenger Type & Occupation')
plt.xticks(fontsize=6)
plt.show()

In [ ]:
print("Acceptance rate of bar coupons by drivers who visit bars more than once a month and have passenger(s) other than kids and have occupations other than farming, fishing, or forestry:", round(acceptance_rate_gt1_nokid_nofff *100, 2),"%")
print("Acceptance rate of bar coupons by all other drivers:", round(acceptance_rate_all_other *100, 2),"%")

Bar coupon acceptance rate of the drivers who visit bars more than once a month and have passengers that are not kids and have occupations other than farming, fishing, or forestry is higher than the other drivers.

6. Compare the acceptance rates between those drivers who:

- go to bars more than once a month, had passengers that were not a kid, and were not widowed *OR*
- go to bars more than once a month and are under the age of 30 *OR*
- go to cheap restaurants more than 4 times a month and income is less than 50K. 



In [ ]:
gt_and_equal1 = ['1~3','4~8','gt8']
passenger_nokid = ['Friend', 'Partner']
age_ls30 = ['below21', '21', '26']
cheap_gt4 = ['4~8', 'gt8']
income_ls50 = ['Less than $12500', '$12500 - $24999', '$25000 - $37499','$37500 - $49999']

group_criteria1 = bar_data.query('Bar in @gt_and_equal1 and passenger in @passenger_nokid and maritalStatus !="Widowed"')
group_criteria2 = bar_data.query('Bar in @gt_and_equal1 and age in @age_ls30')
group_criteria3 = bar_data.query('RestaurantLessThan20 in @cheap_gt4 and income in @income_ls50')

In [ ]:
acceptance_rate1 = group_criteria1['Y'].mean()
acceptance_rate2 = group_criteria2['Y'].mean()
acceptance_rate3 = group_criteria3['Y'].mean()
sns.set(rc={"figure.figsize":(10, 8)})
sns.barplot(x=['Group 1','Group 2', 'Group 3'], y=[acceptance_rate1, acceptance_rate2, acceptance_rate3])
plt.xlabel('Groups with criteria')
plt.ylabel('Acceptance Rate')
plt.title('Bar Coupon Acceptance Rates')
plt.legend(title = 'Group 1: ≥1 Bar visit/month & Non-child passenger & Not widowed \n Group 2: ≥1 Bar visit/month & Under 30 years old \n Group 3: ≥4 Cheap restaurant visit/month & <50K income', title_fontsize=7)
plt.show()

In [ ]:
print("Group 1 Acceptance rate:", round(acceptance_rate1 *100, 2),"%")
print("Group 2 Acceptance rate:", round(acceptance_rate2 *100, 2),"%")
print("Group 3 Acceptance rate:", round(acceptance_rate3 *100, 2),"%")

The comparison of the acceptance rate of the three groups indicates that the first two groups show higher acceptance rates than the third group. However, in order to tell if the criteria can differentiate who would accept the coupon or not, we should compare each group to the groups without respective criteria. 

In [ ]:
all_other_than_criteria1 = bar_data.query('Bar not in @gt_and_equal1 or passenger not in @passenger_nokid or maritalStatus =="Widowed"')
all_other_than_criteria2 = bar_data.query('Bar not in @gt_and_equal1 or age not in @age_ls30')
all_other_than_criteria3 = bar_data.query('RestaurantLessThan20 not in @cheap_gt4 or income not in @income_ls50')

acceptance_rate_all_other1 = all_other_than_criteria1['Y'].mean()
acceptance_rate_all_other2 = all_other_than_criteria2['Y'].mean()
acceptance_rate_all_other3 = all_other_than_criteria3['Y'].mean()

fig, axes = plt.subplots(1, 3, sharey=True, figsize=(12,5))
fig.suptitle('Bar Coupon Acceptance Rates')
axes[0].set_title('≥1 Bar visit/month & Non-child passenger\n& Not widowed vs All other', fontsize=10)
axes[1].set_title('≥1 Bar visit/month & Under 30 years old\nvs All other', fontsize=10)
axes[2].set_title('≥4 Cheap restaurant visit/month & <50K income\nvs All other', fontsize=10)


#Plot Group 1
sns.barplot(ax=axes[0], x=['Group 1','All other'], y=[acceptance_rate1, acceptance_rate_all_other1])

#Plot Group 2
sns.barplot(ax=axes[1], x=['Group 2','All other'], y=[acceptance_rate2, acceptance_rate_all_other2])

# Plot Group 2
sns.barplot(ax=axes[2], x=['Group 3','All other'], y=[acceptance_rate3, acceptance_rate_all_other3]);

Drivers who visit bars frequently show higher acceptance rate whether or not it is combined with other variables. Drivers who visit cheap restaurants frequently and have lower income  show higher acceptance rate than the ones without these criteria, but the rate difference is not very big.

7.  Based on these observations, what do you hypothesize about drivers who accepted the bar coupons?

First hypotheses that we can derive from the data exploration is that the drivers who visit bars frequently are more likely to accept bar coupons.
Second, drivers who have adult passengers might be more likely to accept bar coupons, but we need to explore passenger variable separate from the bar visit variable. 
Third, drivers who visit cheap restaurants frequently and have lower income might be more likely to accept bar coupons
Fourth, age may not be the factor to differentiate who would accept the coupon and who would not. 


### Independent Investigation

Using the bar coupon example as motivation, you are to explore one of the other coupon groups and try to determine the characteristics of passengers who accept the coupons.  

## 1. Create a new DataFrame that contains just the Coffee House coupons.

In [ ]:
#Create coffee_house_data from the cleaned data
coffee_house_data = clean_data.query('coupon == "Coffee House"')
coffee_house_data

## 2. Mapping variables
There are categorial varaiables that show orginal features.
I mapped the ordinal categorical variables to numerical values to preserve the order.

In [ ]:
frequency_mapping = {
    'never': 0,
    'less1': 1,
    '1~3': 2,
    '4~8': 3,
    'gt8': 4
}
age_mapping = {
    'below21': 0,
    '21': 1,
    '26': 2,
    '31': 3,
    '36': 4,
    '41': 5,
    '46': 6,
    '50plus': 7
}
income_mapping = {
    'Less than $12500': 0,
    '$12500 - $24999': 1,
    '$25000 - $37499': 2,
    '$37500 - $49999': 3,
    '$50000 - $62499': 4,
    '$62500 - $74999': 5,
    '$75000 - $87499': 6,
    '$87500 - $99999': 7,
    '$100000 or More': 8
}

cols = ['RestaurantLessThan20', 'Restaurant20To50', 'Bar', 'CoffeeHouse', 'CarryAway']

for col in cols:
    coffee_house_data[col] = coffee_house_data[col].map(frequency_mapping)
    
coffee_house_data['age'] = coffee_house_data['age'].map(age_mapping)
coffee_house_data['income'] = coffee_house_data['income'].map(income_mapping)

#### 'toCoupon_GEQ15min' and 'toCoupon_GEQ25min'

'toCoupon_GEQ15min' and 'toCoupon_GEQ25min' indicate the distance between the drivers' location and the place where they can use the coupons. 
We can combine these two variables into one ordinal variable, 'distance'. 

In [ ]:
coffee_house_data['distance'] = 0 #to assign numerical value to the column
coffee_house_data.loc[coffee_house_data['toCoupon_GEQ15min'] == 0, 'distance'] = 0
coffee_house_data.loc[(coffee_house_data['toCoupon_GEQ15min'] == 1) & (coffee_house_data['toCoupon_GEQ25min'] == 0), 'distance'] = 1
coffee_house_data.loc[coffee_house_data['toCoupon_GEQ25min'] == 1, 'distance'] = 2

## 3. Data Exploration Analysis
I first plotted the histograms of the numercial variables.

In [ ]:
#Select numerical variables and plot histograms
numerical_data = coffee_house_data.select_dtypes('int64', 'float64')
numerical_data.hist(figsize=(12, 12))
plt.suptitle("Histogram of numerical variables")
plt.show()

Then, let's plot the acceptance rate for each ordinal variables.

In [ ]:
# Plot barplots and regression line of the acceptance rates of coffee house coupons for temperature, age, income, CoffeeHouse, CarryAway, RestaurantLessThan20, Restaurant20To50, distance variables.

fig, axes = plt.subplots(2, 4, figsize=(10, 6))
plt.suptitle("Acceptance Rate of Coffee House coupon (Ordinal Variables)")
axes = axes.flatten()

for ax, col in zip(axes, ['temperature', 'age', 'income', 'CoffeeHouse','CarryAway', 'RestaurantLessThan20', 'Restaurant20To50', 'distance']):
    grouped = coffee_house_data.groupby(col)[['Y']].value_counts(normalize=True).unstack().reset_index()
    grouped.columns = grouped.columns.map(str)
    sns.barplot(data=grouped, x=grouped.index, y='1', ax=ax)
    sns.regplot(x=grouped.index, y=grouped['1'], ax=ax)
    ax.set_ylabel('Acceptance Rate')
    ax.set_xlabel(col)


plt.tight_layout();


#### 'temperature', 'RestaurantLessThan20', 'Restaurant20To50'
The barplots of 'temperature', 'RestaurantLessThan20', 'Restaurant20To50' variables show that:
* the acceptance rate consistently increases as the ordinal value becomes bigger,
* and there is positive relationship between these variables and the acceptance rate of coffee house coupon, 

which means that:
* as the temperatrue gets warmer, drivers are more likely use the coffee house coupon,
* and the more frequent the drivers go to restaurants of any prices, the more likely they are to accept the coupon.




#### 'CoffeeHouse'
Aligned with the intuition, the acceptance rate of coffee house coupon increase with frequency of coffee house visit, but it is interesting to see that it does increase until it hits the 3 times a month point, and then decreases a little. 
Also, people who go to restuarants often also uses the coffee coupon more often. 





#### 'distance'
The barplots of 'distance' variable indicates that:
* the acceptance rate consistently decreases as the ordinal value becomes bigger,

which means that:
* as the distance from where they are and to the coffee house location becomes farther, drivers are less likely use the coffee house coupon.

This finding also aligns with common sense. 


#### 'age', 'income', 'CarryAway'
The barplots of 'age', 'income', 'CarryAway' variables show inconsistent acceptance rate meaning ups and downs as the ordinal value increase. 

However, the regression line tells us that there are negative relationships between the acceptance rate and
* 'age'
* 'income'

, and there is a positive relationship between the acceptance rate and
* 'CarryAway'.

While we can look into more statistical evidence whethere or not it is statistically significant in the future analysis, they are meaningful findings for now. 

Below is the inferential statistics (slope and intercept) of the regression lines for each variable.


In [ ]:
reg_data = None
reg_data = pd.DataFrame({'slope': [], 'intercept': []})
index = ['temperature', 'age', 'income', 'CoffeeHouse','CarryAway', 'RestaurantLessThan20', 'Restaurant20To50', 'distance']
for col in ['temperature', 'age', 'income', 'CoffeeHouse','CarryAway', 'RestaurantLessThan20', 'Restaurant20To50', 'distance']:
    grouped = coffee_house_data.groupby(col)[['Y']].value_counts(normalize=True).unstack().reset_index()
    grouped.columns = grouped.columns.map(str)
    pl = sns.regplot(x=grouped.index, y=grouped['1'])
    slope, intercept, r, p, sterr = scipy.stats.linregress(x=pl.get_lines()[0].get_xdata(), y=pl.get_lines()[0].get_ydata())
    reg_data = pd.concat([reg_data, pd.DataFrame({'slope': [slope], 'intercept': [intercept]})])
    plt.close()
    
reg_data.index = index
reg_data

In [ ]:
#Change the variable type to string
coffee_house_data.has_children = coffee_house_data.has_children.astype(str)
coffee_house_data.direction_same = coffee_house_data.direction_same.astype(str)

#Visualize barplots
fig, axes = plt.subplots(4,2, figsize=(12, 12))
plt.suptitle("Acceptance Rate of Coffee House coupon (Categorical Variables)")
axes = axes.flatten()

for ax, col in zip(axes, ['destination', 'passenger', 'weather', 'expiration','gender', 'maritalStatus', 'has_children', 'direction_same']):
    grouped = coffee_house_data.groupby(col)[['Y']].value_counts(normalize=True).unstack()
    
    sns.barplot(data=grouped, x=grouped.index, y=1, ax=ax)
    ax.set_ylabel('Acceptance Rate')
    
plt.tight_layout();

The barplots of the acceptance rate for different categorical variables indicate that drivers tend to accept coupons more when:
* they don't have any urgent places to go,
* they have adult passengers,
* it is not snowing,
* the coupon expiration period is longer,
* and they are not widowed.

Drivers' gender, whether they have children or not, and whether they are going to the same direction as the location of the coffee house only show small difference in the percentage. 

## 4. Conclusion

Based on the data exploration of the Coffee House coupons, the following hypotheses can be derived regarding the characteristics of drivers who accepted the coupons:

* Temperature: Drivers are more likely to use the coffee house coupon as the temperature gets warmer.
* Age: Younger drivers are more likely to accept the coffee house coupon compared to older drivers.
* Coffee House Visit Frequency: As the frequency of visiting coffee houses increases, the acceptance rate of coffee house coupons also increases until it reaches a peak at around 3 visits per month and then slightly decreases.
* Restaurant Visit Frequency: Drivers who visit restaurants more frequently, regardless of the price range, are more likely to accept the coffee house coupon.
* Distance: The acceptance rate of the coffee house coupon decreases as the distance from the driver's location to the coffee house increases.
* Carry Away: Drivers who frequently purchase take-away food are more likely to accept the coffee house coupon.
* Other Variables: Some categorical variables such as destination, passenger type, weather, expiration, marital status show affect the acceptance rate of the coffee house coupon.
* income, gender, presence of children, and same direction do not show consistent patterns or strong relationships.

These hypotheses provide insights into the characteristics of drivers who are more likely to accept Coffee House coupons. These insights can be utilized when deciding target demographic/environment to distribute the coupons.

However, further statistical analysis and hypothesis testing would be required to validate these findings and determine the significance of the relationships observed.
